# Data Wrangling

- `pandas`
- Where to find data?
   - Web Scraping & APIs
   
   
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/COGS108/Lectures-Fa24/blob/main/03_pandas.ipynb)





<center>
<img src="https://raw.githubusercontent.com/COGS108/Lectures-Wi22/main/02_python/img/pandas.png" alt="pandas" width="600px">
</center>



Pandas is Python library for managing heterogenous data.

At its core, Pandas is used for the **DataFrame** object, which is:
- a data structure for labeled rows and columns of data
- associated methods and utilities for working with data.
- each column contains a `pandas` **Series**

# After this

When you are done with this exercise/lecture... the real learning can begin!

One of your best tools is https://pandastutor.com

Another is the documentation for pandas including https://pandas.pydata.org/docs/user_guide/10min.html

## Setup

In [1]:
# Import standard libraries
%matplotlib inline
import pandas as pd
import numpy as np

In [2]:
# reminder about tab completion and contextual help

## Loading Data

In [3]:
# Load a csv file of data
df = pd.read_csv('https://raw.githubusercontent.com/COGS108/Lectures-Fa24/main/data/my_data.csv')
print('total elements:',df.size,'\nshape of table',df.shape)

total elements: 1200 
shape of table (200, 6)


In [4]:
# Check out a few rows or last few rows of the dataframe using head() or tail()
df

id first_name last_name  age  score     value
0      295     Andrea     Clark   46     -1  24547.87
1      620       Bill     Woods   46    492  46713.90
2      891  Alexander  Jacobson   48    489  32071.74
3      914    Derrick   Bradley   52     -1  30650.48
4     1736    Allison    Thomas   44     -1   9553.12
..     ...        ...       ...  ...    ...       ...
195  97441     Krista     Ortiz   34     -1  24074.79
196  97728       Anna  Chambers   37    598      0.00
197  98115   Jennifer     Pitts   29    606   6876.75
198  98284   Brittany   Jenkins   34    665  43525.88
199  98366    Katelyn     Brown   45    501  29668.38

[200 rows x 6 columns]

Pandas DataFrame:
- Index for each row
- Column name for each column (Series)
- Stores heterogenous types

## Slicing

In [5]:
# Slicing (Indexing): select a Series (column) using its name

df.loc[0:11]

id first_name last_name  age  score      value
0    295     Andrea     Clark   46     -1   24547.87
1    620       Bill     Woods   46    492   46713.90
2    891  Alexander  Jacobson   48    489   32071.74
3    914    Derrick   Bradley   52     -1   30650.48
4   1736    Allison    Thomas   44     -1    9553.12
5   2049    Stephen  Williams   57    333  138936.92
6   2241      Malik      Wood   46     -1   10804.47
7   2607      Amber    Garcia   50    536    9367.27
8   2635      David   Coleman   68    351   66035.28
9   3585       Eric    Atkins   56    582  103977.32
10  4199     Justin   Johnson   59    500   34938.08
11  6739      Donna    Barnes   48    500  130915.20

In [6]:
df.loc[3:8, :]

id first_name last_name  age  score      value
3   914    Derrick   Bradley   52     -1   30650.48
4  1736    Allison    Thomas   44     -1    9553.12
5  2049    Stephen  Williams   57    333  138936.92
6  2241      Malik      Wood   46     -1   10804.47
7  2607      Amber    Garcia   50    536    9367.27
8  2635      David   Coleman   68    351   66035.28

In [7]:
df.loc[:,'last_name']

0         Clark
1         Woods
2      Jacobson
3       Bradley
4        Thomas
         ...   
195       Ortiz
196    Chambers
197       Pitts
198     Jenkins
199       Brown
Name: last_name, Length: 200, dtype: object

In [15]:
df.loc[3:8,'last_name']
# We aren't slicing by index value, its the index name which just happens to be numeric in this case which is why we don't have to end with 9 to include index 8

3     Bradley
4      Thomas
5    Williams
6        Wood
7      Garcia
8     Coleman
Name: last_name, dtype: object

In [19]:
#df.set_index('last_name')

df.iloc[3:8] #iloc is positional base

id first_name last_name  age  score      value
3   914    Derrick   Bradley   52     -1   30650.48
4  1736    Allison    Thomas   44     -1    9553.12
5  2049    Stephen  Williams   57    333  138936.92
6  2241      Malik      Wood   46     -1   10804.47
7  2607      Amber    Garcia   50    536    9367.27

In [20]:
# Slicing: select a row & column with 'loc'
df.loc[4, 'age']

44

In [21]:
df['age'] > 10

0      True
1      True
2      True
3      True
4      True
       ... 
195    True
196    True
197    True
198    True
199    True
Name: age, Length: 200, dtype: bool

In [22]:
df.loc[:,'age']

0      46
1      46
2      48
3      52
4      44
       ..
195    34
196    37
197    29
198    34
199    45
Name: age, Length: 200, dtype: int64

#### Question #1

What would be the output of `df['age'] > 10`?

- A) subset of `df` including only rows of individuals older than 10
- **B) a Boolean with `True` for rows where age is greater than 10 and `False` otherwise**
- C) `id`s of rows where observations are greater than 10
- D) an error
- E) I'm super lost

In [25]:
# to get subset
df.loc[ (df['age'] > 60), ['first_name','last_name'] ]
# list of the names of people over 60 years old

first_name last_name
8        David   Coleman
22      Daniel       Key
39     Kenneth   Johnson
51      Dennis     Mason
53       James   Kennedy
54       Jason   Bentley
85     Michael     Cowan
89     Maxwell   Mcbride
94       Bryan   Barnett
109      Jared     Brown
125   Madeline    Rivera
158     Brenda       Cox
177     Robert     Woods

No charts were generated by quickchart


In [27]:
df.query('score < 0')

id first_name  last_name  age  score     value
0      295     Andrea      Clark   46     -1  24547.87
3      914    Derrick    Bradley   52     -1  30650.48
4     1736    Allison     Thomas   44     -1   9553.12
6     2241      Malik       Wood   46     -1  10804.47
19   12250      Erica      Adams   41     -1   2126.22
24   13806     Ashley      Jones   57     -1  80325.66
27   15137     Andrea    Simpson   55     -1  12155.47
35   19129      Brian     Travis   46     -1  49955.06
36   19557    Sabrina      Simon   44     -1  14022.29
39   22584    Kenneth    Johnson   65     -1  50174.14
41   24650  Stephanie      Hayes   55     -1  15625.50
42   24779      Scott      Reyes   54     -1  65163.92
57   31457       John    Shields   51     -1  25184.43
61   33863      Brian     Murphy   45     -1  20802.53
70   37836    Kristin       King   34     -1  24409.19
72   38892   Kimberly  Jefferson   60     -1  28155.47
77   42140      Aaron       Hart   37     -1  16873.92
82   45647    Melissa     Bailey   46     -1   4104.35
92   51909    Michael    Brennan   45     -1  12838.06
96   53812    Melissa     Walton   51     -1      0.00
101  54949  Christine      Myers   45     -1  15230.16
105  57077      Jason      Price   32     -1  24557.68
122  64731     Sheila  Zimmerman   32     -1   9726.45
153  82183     Tamara    Jenkins   46     -1  23914.62
157  84203     Warren     Wilson   34     -1  17172.97
158  84475     Brenda        Cox   64     -1  91349.57
163  84965     Robert      Black   59     -1  13873.11
168  87122    Douglas      Hogan   49     -1  21598.29
182  91893     Kelsey     Martin   52     -1  40206.46
185  92298   Jennifer      Marks   39     -1  36443.84
190  94730       Lisa  Dominguez   22     -1       NaN
193  96293    Michael   Phillips   32     -1   9159.90
195  97441     Krista      Ortiz   34     -1  24074.79

In [28]:
df.query('first_name == "Andrea"')

id first_name last_name  age  score     value
0     295     Andrea     Clark   46     -1  24547.87
27  15137     Andrea   Simpson   55     -1  12155.47

In [32]:
# NOTE: .loc uses the index, whatever it is
# sometimes the index is NOT numeric!
df.set_index('age') #.loc[52]

df.set_index('first_name').loc['Andrea']


id last_name  age  score     value
first_name                                       
Andrea        295     Clark   46     -1  24547.87
Andrea      15137   Simpson   55     -1  12155.47

In [30]:
(
    df
    .set_index('first_name')
    .loc['Andrea']
)

id last_name  age  score     value
first_name                                       
Andrea        295     Clark   46     -1  24547.87
Andrea      15137   Simpson   55     -1  12155.47

## Checking out the DataFrame

In [33]:
# Check how large our dataframe is
df.shape

(200, 6)

In [34]:
# Check what columns we have in our DataFrame
df.columns

Index(['id', 'first_name', 'last_name', 'age', 'score', 'value'], dtype='object')

In [35]:
df.set_index('first_name').index

Index(['Andrea', 'Bill', 'Alexander', 'Derrick', 'Allison', 'Stephen', 'Malik',
       'Amber', 'David', 'Eric',
       ...
       'Lisa', 'Julia', 'Wayne', 'Michael', 'Sherri', 'Krista', 'Anna',
       'Jennifer', 'Brittany', 'Katelyn'],
      dtype='object', name='first_name', length=200)

In [36]:
# Check the datatypes of our variables
df.dtypes

id              int64
first_name     object
last_name      object
age             int64
score           int64
value         float64
dtype: object

## Exploring the data

- quantitative (numbers)
- qualitative (categorical)
- basic descriptive statistics

In [37]:
# Checking categorical data
df['first_name'].value_counts()
# try adding this: .iloc[:20]

first_name
David       6
Michael     5
Eric        4
Charles     4
James       4
           ..
Alison      1
Andrew      1
Vanessa     1
Samantha    1
Katelyn     1
Name: count, Length: 134, dtype: int64

In [38]:
# Check a particular descriptive statistic, like mean(), median(), mode(), std(), var(), etc
df[['id','age', 'value']].median()

id       54643.50
age         46.00
value    17976.51
dtype: float64

In [39]:
# Describe a particular column
df['value'].describe()

count       189.000000
mean      28730.336296
std       32493.945741
min           0.000000
25%        9593.030000
50%       17976.510000
75%       33163.310000
max      204999.960000
Name: value, dtype: float64

In [40]:
# Get descriptive statistics of all numerical columns
df.describe()

id         age       score          value
count    200.000000  200.000000  200.000000     189.000000
mean   52929.150000   46.020000  416.595000   28730.336296
std    29414.298899   10.028582  237.176674   32493.945741
min      295.000000   14.000000   -1.000000       0.000000
25%    26709.500000   39.000000  288.750000    9593.030000
50%    54643.500000   46.000000  463.500000   17976.510000
75%    80840.750000   53.000000  596.500000   33163.310000
max    98366.000000   69.000000  942.000000  204999.960000

#### Question #2

What's the average (mean) age of the individuals in this dataset?

- A) 14
- **B) 46**
- C) 28730
- D) NA
- E) I'm super lost/unsure

In [44]:
# there are two commands that could give you the correct answer...
# try it out here!
df['age'].mean()

46.02

## `pandas`: Common Manipulations

You'll want to be *very* familiar with a few common data manipulations when wrangling data, each of which is described below:

Manipulation | Description
-------|------------
**select** | select which columns to include in dataset
**filter** | filter dataset to only include specified rows
**mutate** | add a new column based on values in other columns
**groupby** | group values to apply a function within the specified groups
**summarize** | calculate specified summary metric of a specified variable
**arrange** | sort rows ascending or descending order of a specified column
**merge** | join separate datasets into a single dataset based on a common column



In [42]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Selecting & Dropping Columns

- include subset of columns of larger data frame

In [45]:
df.head()

id first_name last_name  age  score     value
0   295     Andrea     Clark   46     -1  24547.87
1   620       Bill     Woods   46    492  46713.90
2   891  Alexander  Jacobson   48    489  32071.74
3   914    Derrick   Bradley   52     -1  30650.48
4  1736    Allison    Thomas   44     -1   9553.12

In [47]:
# specify which columns to include
select_df = df[['id', 'age', 'score', 'value']] #makes subset named select_df
select_df.head()

id  age  score     value
0   295   46     -1  24547.87
1   620   46    492  46713.90
2   891   48    489  32071.74
3   914   52     -1  30650.48
4  1736   44     -1   9553.12

In [ ]:
# Drop rows we don't want
df.drop(labels=[0,2,6], axis='rows')

In [ ]:
# Check out the DataFrame after dropping, what went wrong and how do we fix it?
df.head()

## Filtering Data (slicing)

- include a subset (slice) of rows from larger data frame

In [ ]:
# Check if we have any data from people below the age of 18
sum(df['age'] < 18)

In [ ]:
df.loc[(df['age'] < 18)]

In [ ]:
# Select only participants who are 18 or older  AND who have a score of -1
df_new = df[ (df['age'] >= 18)] # & (df['score']==-1)]
df_new

## Missing Data (NaNs)

In [ ]:
# Check for missing values
df['value'].hasnans

In [ ]:
df.isnull?

In [ ]:
# note in class
# can operate on entire dataframe
df.isnull()

In [ ]:
# Check for null values by row
df.isnull().sum(axis='rows')

In [ ]:
# Have a look at the missing values
df[df['value'].isnull()]
# can also use .isna() or isnan()!

## Dealing with Missing Data - NaNs

In [ ]:
# Dealing with null values: Drop rows with missing data
print('before dropping null rows',df.shape)
no_na_df = df.dropna()
print('new shape', no_na_df.shape)

In [ ]:
# or you can fill in a value for those missing values!
df.fillna?
# df.fillna(100) or df.fillna( df['value'].mean() ) or just df.fillna()

## Finding Missing Data - Bad Values

In [ ]:
# Check for the properties of specific columns
df['score'].describe()

In [ ]:
df[df.score<0].shape

In [ ]:
# Check the plot of the data for score to see the distribution
df['score'].plot(kind='hist', bins=25);

In [ ]:
df.hist();

## Dealing with Missing Data - Bad Values

In [ ]:
# Look for how many values have a -1 value in 'score'
sum(df['score'] == -1)

In [ ]:
# Drop any row with -1 value in 'score'
df = df[df['score'] != -1]
df.shape

## Creating new columns (mutating)

- `assign` can be very helpful in adding a new column
- lambda functions can be used to carry out calculations

In [ ]:
# convert age in years to age in (approximate) days
df = df.assign(age_days = df['age'] * 365)
df.head()

In [ ]:
df['age_months'] = df['age'] * 12
df.head()

## Grouping & summarizing

- group by a particular variable
- calculate summary statistics/metrics within group

In [ ]:
df.first_name.value_counts()

In [ ]:
# caclculate average within each age
df.groupby('first_name').age.mean()

## Sorting Rows (arrange)

- specify order in which to display rows

In [ ]:
df.head()

In [ ]:
# sort by values in age
df = df.sort_values(by = ['age'],ascending=True)
df.head()

## Combining datasets
![image of join operations](https://raw.githubusercontent.com/COGS108/Lectures-Fa24/main/img/join.png)

In [ ]:
## Create two DataFrames
left = pd.DataFrame({'key': ['A', 'B', 'C', 'D'], 'value': np.random.randn(4)})
right = pd.DataFrame({'key': ['B', 'D', 'E', 'F'], 'value': np.random.randn(4)})

In [ ]:
left

In [ ]:
right

In [ ]:
left.merge?

In [ ]:
left.merge(right, on='key', )

In [ ]:
# inner merge by default
pd.merge(left, right, on='key')

In [ ]:
# same as above
pd.merge(left, right, on='key', how='inner')

In [ ]:
# right merge
pd.merge(left, right, on='key', how='right')

In [ ]:
# left merge
pd.merge(left, right, on='key', how='left')

In [ ]:
# outer join
pd.merge(left, right, on='key', how='outer')

In [ ]:
pd.merge(left, right, on='key', how='outer').mean()

Let's do this as a check in!
## https://forms.gle/EpNS7BYhfAxvBM4X7

#### Question #3

If table A had 5 rows and table B had 5 rows and 3 of those rows in each table were from the same observations present in the *other* table, how many rows would be present if an **inner merge** were carried out?

- A) 3
- B) 5
- C) 10
- D) 13
- E) Totally unsure

#### Question #4

If table A had 5 rows and table B had 5 rows and 3 of those rows in each table were from the same observations present in the *other* table, how many rows would be present if a **left merge** were carried out?

- A) 3
- B) 5
- C) 10
- D) 13
- E) Totally unsure

## Visualizing Data

- We'll have a whole lecture (or two) on visualization
- For now, we'll just look at one uniquely-pandas approach

In [ ]:
# Plot all numerical columns, and their interactions
pd.plotting.scatter_matrix(df[['age', 'score', 'value']], figsize=[12, 12], marker=12);

## Application Program Interface (APIs)

- APIs are basically a way for software to talk to software
    - They are an interface into an application / website / database designed for computers / software.

Notes on APIs:
- Follow API guidelines!
- These guidelines typically specify the number / rate / size of requests

## Github API

You can access the github api with the following API. Just added specifiers for what you are looking for.

https://api.github.com/

For example, the following URL will search for the user 'ShanEllis':

https://api.github.com/users/shanellis




## Requesting Web Pages from Python

In [ ]:
# The requests module allows you to send URL requests from python
import requests
from bs4 import BeautifulSoup

In [ ]:
# Request data from the Github API on a particular user
page = requests.get('https://api.github.com/users/jasongfleischer')

In [ ]:
# The content we get back is a messily organized json file
page.content

#### Question #5

What type/format of output is this?

- A) CSV
- B) XML
- C) JSON
- D) API
- E) I'm super lost

In [ ]:
# We can read in the json data with pandas
git_data = pd.read_json(page.content, typ='series')

In [ ]:
# Check out the pandas series object full of data
git_data

### Authorized Access - OAuth

Open Authorization is a protocol to authorize access (of a user / application) to an API.

OAuth provides a secure way to 'log-in' without using account names and passwords.

It is effectively a set of keys, and passwords you can use to access APIs.

## Web Scraping vs. APIs

Web scraping and APIs are different approaches:

- APIs are an interface to interact with an application, designed for programmatic use
    - They allow systematic, controlled access to (for example) and applications database
    - They typically return structured (friendly) data

- Web scraping (typically) involves navigating through the internet, programmatically following an architecture built for humans
    - This can be hard to systematize, being dependent on the idiosyncracies of a web page, at the time you request it
    - This typically returns relatively unstructured data
    - This entails much more wrangling of the data

## Where to Find Data?

* [Awesome Public Datasets](https://github.com/awesomedata/awesome-public-datasets/blob/master/README.rst)
* [Data.gov](https://catalog.data.gov/dataset)
* [Data Is Plural](https://docs.google.com/spreadsheets/d/1wZhPLMCHKJvwOkP4juclhjFgqIY8fQFMemwKL2c64vk/edit#gid=0)
* [UCSD Datasets](https://ucsd.libguides.com/data-statistics/home)
* [Datasets | Deep Learning](http://deeplearning.net/datasets/)
* [Stanford | Social Science Data Collection](https://data.stanford.edu/)
* [Eviction Lab (email required)](https://evictionlab.org/get-the-data/)
* [San Diego Data](https://data.sandiego.gov/)
* [US Census](https://www.census.gov/)
* [Open Climate Data](http://openclimatedata.net/)
* [Data and Story Library](https://dasl.datadescription.com/datafiles/)
* [UCSD behavioral mobile data](http://extrasensory.ucsd.edu/)
* [Kaggle](https://www.kaggle.com/)
* [FiveThirtyEight](https://data.fivethirtyeight.com/)
* [data.world](https://data.world/)
* [Free Datasets - R and Data Mining ](http://www.rdatamining.com/resources/data)
* [Data Sources for Cool Data Science Projects](https://blog.thedataincubator.com/2014/10/data-sources-for-cool-data-science-projects-part-1/)

## Notes on Working with Data

### Data Science is Ad-Hoc

- It is part of the job description to put things together that were not designed to go together.
- We do not have universal solutions, but haphazard, idiosyncratic systems, for data collection, storage and analysis.
- Data is everywhere. But relatively little of it was collected *as data*.

### Data Collection, Curation, and Storage are Difficult

- It can be difficult to choose broadly useful standards
- Take time to think about your data, and how you will load, store, organize and save it

### Data is Inherently Noisy

- We live in a messy, noisy, world, with messy, noisy, people, using messy, noisy instruments.
- There is no perfect data.
    - There is better / or worse data, given the context.

### Different Objectives

- Humans and computers are different.
- We interact with '*data*' in different ways.
- This underlies many aspects of data wrangling
    - The 'friendliness' of data types / files
    - The difference between web scraping and APIs
    - A disconnect between data in the real world, and data we want to use

## So... What to do?

- Think about how your data are stored & its structure?
- Look at your data before you anayze it
    - are there missing values?
    - outlier values?
- Are your data trustworthy?
    - source?
    - how was it generated?

## Specific Recommendations

- Prioritize using well structured, common, open file types
    - Take advantage of existing tools to deal with these files (numpy, pandas, etc.)

- Look into, and then follow, common conventions
    - Minimize custom objects, workflows and data files
- Look for APIs. Ask if they are available.
    - Acknowledge that web scraping and/or wrangling unstructured data are complex / long tasks

- Think about data flow from the beginning. Organize your data pipeline, consider the 'wrangling' aspects throughout
    - Set yourself up with well organized, labelled approach to your data
    - Think about when and how you might want/need to save out intermediate results.